### DISCUSSION TITLE
Matching the shape of y and X (e.g. when introducing missing values due to differencing)

### DISCUSSION THREAD

**Discussion by hliebert** [alan-turing-institute/sktime#4882] [Q&A] [answered]
In the mock example below, using `Differencer()` with `na_handling="drop_na"` triggers `ValueError: Found input variables with inconsistent numbers of samples: [16, 15]`. Setting `na_handling="keep_na"` triggers `ValueError: Input y contains NaN`.

I've mostly opted for combining  `keep_na` with imputing missings rather than using the default `fill_zero` (I was a bit puzzled about this being the default).

Is there an easy way of using the `drop_na` variant and ensure the dimensions of y and X match?

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import MinMaxScaler
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster, make_reduction
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.feature_selection import FeatureSelection
from sktime.transformations.series.impute import Imputer

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 3), is_relative=True)

random_forest = RandomForestRegressor()
reduction_forecaster = make_reduction(random_forest, window_length=3, strategy="direct", windows_identical=False)

pipe = ForecastingPipeline(
    [
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        ("feature_selector", FeatureSelection()),
        ("reduction_forecaster", reduction_forecaster),
    ]
)

forecaster = TransformedTargetForecaster(
    [
        ("log_transformer", LogTransformer()),
        ("first_differencer", Differencer(lags=1, na_handling="drop_na")),  # keep_na triggers ValueError: Input y contains NaN.
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        # ("imputer", Imputer(method="drift")),  # o/w cut X to same length?
        ("forecasting_pipeline", pipe),
    ]
)

forecaster.fit(y, X, fh=horizon)
forecaster.predict(X=X, fh=horizon)

```

**Comment by fkiraly** [comment, marked as answer]
Commenting separately on "is" and "should".

In the "is" state - knocking on wood here - it might work if you use the `IndexSubset` transform from `transformations.series.subset`.
You have to use it in the "`y`-pipe" (the `ForecastingPipeline` `pipe`) so it subsets `y` to the index of `X`.

You may also have to nest the two pipelines the right way round, but from a superficial glance it looks like it is the way round it should be, since the `X`-transformations are applied before the `y`-transformations, so `IndexSubset` as a `y`-transformation would already see the smaller index subset of `X`.

I can't guarantee that it works and it also seems a bit counterintuitive imho, so would appreciate usage feedback.

Regarding "should": perhaps, if the forecaster has the tag `X-y-must-have-same-index`, it should just subset `X` to the `y`-index automatically if that's possible? (or can we see a problem with that?)

**Reply by fkiraly** [reply]
oh wait, I confused myself. It's actually the other way round. `y` is shorter than `X`.
So at current state, you have to `IndexSubset` in the `TransformedTargetForecaster`, and ensure it gets applied after the `Differencer`, which would require the nesting to be the other way round.

(in the "should" case, it wouldn't matter, would it? thinking in theory to see whether that's a good solution)

**Reply by hliebert** [reply]
Thanks for the quick reply.  If I understand correctly, `ForecastingPipeline `applies transformations to X only (and `TransformedTargetForecaster `to y). In the example above, transformations to y occur before transformations to X. 

In any case, adding `IndexSubset(index_treatment="remove"))` to `ForecastingPipeline()` works in the above case. 

Regarding the "should" case, your proposal sounds sensible to me.

**Reply by fkiraly** [reply]
> If I understand correctly, `ForecastingPipeline` applies transformations to `X` only (and `TransformedTargetForecaster` to `y`). 

Yes - although the transformers in `ForecastingPipeline` get to see the forecaster's `y` as `y` in `transform`, and the transformers in `TransformedTargetForecaster` get to see the forecaster's `X` as `y` in `transform`, that's why `IndexSubset` should work.

**Reply by fkiraly** [reply]
Could you perhaps post the code that works? Just to settle my confusion here, would be much appreciated as it helps with work on the "should" case.

**Reply by hliebert** [reply]
Sure. The following works.

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import MinMaxScaler
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster, make_reduction
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.feature_selection import FeatureSelection
from sktime.transformations.series.impute import Imputer
from sktime.transformations.series.subset import IndexSubset

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 3), is_relative=True)

random_forest = RandomForestRegressor()
reduction_forecaster = make_reduction(random_forest, window_length=3, strategy="direct", windows_identical=False)

pipe = ForecastingPipeline(
    [
        ("index_subset", IndexSubset(index_treatment="remove")),
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        ("feature_selector", FeatureSelection()),
        ("reduction_forecaster", reduction_forecaster),
    ]
)

forecaster = TransformedTargetForecaster(
    [
        ("log_transformer", LogTransformer()),
        ("first_differencer", Differencer(lags=1, na_handling="drop_na")),  # keep_na triggers ValueError: Input y contains NaN.
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        ("forecasting_pipeline", pipe),
    ]
)

forecaster.fit(y, X, fh=horizon)
forecaster.predict(X=X, fh=horizon)
```
